packages/electron-builder/src/codeSign.ts

```diff
@@ -221,16 +221,12 @@ export async function findIdentity(certType: CertType, qualifier?: string | null
     for (const prefix of appleCertificatePrefixes) {
       checkPrefix(identity, prefix)
     }
-    const result = await _findIdentity(certType, identity, keychain)
-    if (result == null) {
-      throw new Error(`Identity name "${identity}" is specified, but no valid identity with this name in the keychain`)
-    }
-    return result
+    return await _findIdentity(certType, identity, keychain)
```

Comment:
[macPackager.ts](https://github.com/electron-userland/electron-builder/blob/master/packages/electron-builder/src/macPackager.ts#L149)
```
    let name = await findIdentity(isMas ? "3rd Party Mac Developer Application" : "Developer ID Application", isMas ? masQualifier : this.platformSpecificBuildOptions.identity, keychainName)
    if (name == null) {
      if (!isMas) {
        name = await findIdentity("Mac Developer", this.platformSpecificBuildOptions.identity, keychainName)
        if (name != null) {
          warn("Mac Developer is used to sign app — it is only for development and testing, not for production")
        }
      }

      if (name == null) {
        const message = process.env.CSC_IDENTITY_AUTO_DISCOVERY === "false" ?
          `App is not signed: env CSC_IDENTITY_AUTO_DISCOVERY is set to false` :
          `App is not signed: cannot find valid ${isMas ? '"3rd Party Mac Developer Application" identity' : `"Developer ID Application" identity or custom non-Apple code signing certificate`}, see https://github.com/electron-userland/electron-builder/wiki/Code-Signing`
        if (isMas || this.platformSpecificBuildOptions.forceCodeSigning) {
          throw new Error(message)
        }
        else {
          warn(message)
          return
        }
      }
    }
```

As for the current logic in `macPackager.ts`, when provided with an identity string, it first try to get it as a "3rd Party Mac Developer Installer"/"Developer ID Installer"; If returns null, it will do an additional identity query for Darwin as 'Mac Developer'; If returns null again, it will decide to either throw an error or display a warning message depending on the build options(`isMas || this.platformSpecificBuildOptions.forceCodeSigning`);

So here in `findIdentity()` I don't see why it should throw an error when failed to get identity. It can just return `string|null` denoting whether it find the requested identity or not, and let `macPackager.ts` do the result checking part. `findIdentity()` should not throw error before every options are tried in `macPackager.ts`.